src/include/duckdb/common/types/vector.hpp

```diff
@@ -38,13 +39,32 @@ struct UnifiedVectorFormat {
 	data_ptr_t data;
 	ValidityMask validity;
 	SelectionVector owned_sel;
+	PhysicalType physical_type;
 
 	template <class T>
-	static inline const T *GetData(const UnifiedVectorFormat &format) {
+	void VerifyVectorType() const {
+#ifdef DUCKDB_DEBUG_NO_SAFETY
+		D_ASSERT(StorageTypeCompatible<T>(physical_type));
+#else
+		if (!StorageTypeCompatible<T>(physical_type)) {
+			throw InternalException("Expected unified vector format of type %s, but found type %s", GetTypeId<T>(),
+			                        physical_type);
+		}
+#endif
+	}
+
+	template <class T>
+	static inline const T *GetDataUnsafe(const UnifiedVectorFormat &format) {
 		return reinterpret_cast<const T *>(format.data);
 	}
 	template <class T>
+	static inline const T *GetData(const UnifiedVectorFormat &format) {
+		format.VerifyVectorType<T>();
```

Comment:
```suggestion
#ifndef DISABLE_SAFE_VECTORS
		format.VerifyVectorType<T>();
#endif
```
or equivalent (also 5 lines below)?

That would allow (with some extra variable passing) to then plug this in "Regression Tests between safe and unsafe builds" that are run like:
```
         make

         cd unsafe
         UNSAFE_NUMERIC_CAST=1 DISABLE_MEMORY_SAFETY=1 make
```

Reply:
This already happens in `VerifyVectorType` directly, no?

Reply:
```cpp
	void VerifyVectorType() const {
#ifdef DUCKDB_DEBUG_NO_SAFETY
		D_ASSERT(StorageTypeCompatible<T>(physical_type));
#else
		if (!StorageTypeCompatible<T>(physical_type)) {
			throw InternalException("Expected unified vector format of type %s, but found type %s", GetTypeId<T>(),
			                        physical_type);
		}
#endif
	}
```